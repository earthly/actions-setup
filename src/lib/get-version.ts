import { Octokit } from "@octokit/core";
import { Endpoints } from "@octokit/types";
import { maxSatisfying, gte } from "semver";
import { paginateRest } from "@octokit/plugin-paginate-rest";
import * as core from "@actions/core";

type ReleaseResponse =
  Endpoints["GET /repos/{owner}/{repo}/releases"]["response"]["data"][0];

export async function getVersionObject(
  range: string,
  prerelease: boolean,
): Promise<ReleaseResponse> {
  const MyOctokit = Octokit.plugin(paginateRest);
  const octokit = new MyOctokit({
    auth:
      core.getInput("github-token") || process.env.GITHUB_TOKEN || undefined,
  });
  const versions = (
    await octokit.paginate("GET /repos/{owner}/{repo}/releases", {
      owner: "idodod",
      repo: "earthly",
      per_page: 100,
    })
  ).filter(release => {
    return (prerelease || !release.prerelease) && release.assets?.length > 0;
  }).reduce((acc, cur) => {
    // remove 'v' from tag name
    const tag = cur.tag_name.substring(1);

    acc[tag] = cur;

    return acc;
  }, {} as Record<string, ReleaseResponse>);

  if (range == "latest") {
    const latest = Object.keys(versions).reduce((prev, cur) => {
      return gte(cur, prev) ? cur : prev;
    });
    invariant(latest, "expect a latest version to exists");
    return versions[latest];
  }

  const resp = maxSatisfying(Object.keys(versions), range);
  if (resp === null) {
    throw new Error(
      "Could not find a version that satisfied the version range"
    );
  }

  const ver = versions[resp];
  if (!ver) {
    throw new Error(
      "Could not find a version that satisfied the version range"
    );
  }

  return ver;
}

/* eslint @typescript-eslint/explicit-module-boundary-types: 0 */
export function invariant(
  condition: unknown,
  message?: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}
