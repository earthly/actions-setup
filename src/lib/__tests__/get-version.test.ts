import { getVersionObject } from "../get-version";
import * as semver from 'semver';

// The latest version since this test was last changed
// Feel free to update it if earthly has been updated
const latest = '0.6.23';

describe("get-version", () => {
  describe('latest range versions', () => {
    it.each(["latest", "^0", "0.*.*"] as const)("should match %s versions", async (ver) => {
      const v = await getVersionObject(ver);
      expect(semver.gte(v.tag_name, latest));
    });
  });
  describe("range versions", () => {
    it.each([{spec: "0.4.*", gte: "0.4.0"}, {spec: "v0.4.*", gte: "0.4.0"}, {spec: "0.6.1", eq: '0.6.1'}, {spec: "v0.6.0", eq: "0.6.1"}] as const)("should match %s versions", async (test) => {
      const v = await getVersionObject(test.spec);
      if (test.gte) expect(semver.gte(v.tag_name, test.gte));
      if (test.eq) expect(semver.eq(v.tag_name, test.eq));
    });
  });
});
