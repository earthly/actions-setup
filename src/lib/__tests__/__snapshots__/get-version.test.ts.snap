// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-version range versions should match ^0 versions 1`] = `"v0.6.23"`;

exports[`get-version range versions should match 0.*.* versions 1`] = `"v0.6.23"`;

exports[`get-version range versions should match 0.4.* versions 1`] = `"v0.4.6"`;

exports[`get-version range versions should match 0.6.1 versions 1`] = `"v0.6.1"`;

exports[`get-version range versions should match 0.6.1 versions 2`] = `"v0.6.1"`;

exports[`get-version range versions should match latest versions 1`] = `"v0.6.23"`;

exports[`get-version range versions should match v0.4.* versions 1`] = `"v0.4.6"`;

exports[`get-version range versions should match v0.6.0 versions 1`] = `"v0.6.0"`;
